### ISSUE TITLE
not able to extract pubkey from BEID (Belgium ID card) with ssh 

### ISSUE THREAD

**Issue by francois4224** [OpenSC/OpenSC#1430]
I'm able to extract the authentication ssh key with  `pkcs15-tool --read-ssh 2` but when I try to ssh with onepin-opensc-pkcs11.so I have this: 

```
ssh -vv -I "/Library/OpenSC/lib/pkcs11/onepin-opensc-pkcs11.so" me@myserver
...
debug1: Connection established.
debug1: provider /Library/OpenSC/lib/pkcs11/onepin-opensc-pkcs11.so: manufacturerID <OpenSC Project> cryptokiVersion 2.20 libraryDescription <OpenSC smartcard framework> libraryVersion 0.18
debug1: provider /Library/OpenSC/lib/pkcs11/onepin-opensc-pkcs11.so slot 0: label <Basic PIN (BELPIC)> manufacturerID <(unknown)> model <PKCS#15> serial <293E15001292626> flags 0x40d
debug1: pkcs11_add_provider: provider /Library/OpenSC/lib/pkcs11/onepin-opensc-pkcs11.so returned no keys
...
```

I'm able to login with a Yubikey Neo and with a Feitian key in the same way.

**Comment by francois4224** [comment]
and I use OpenSC 0.18.0 on MacOS

**Comment by frankmorgner** [comment]
Could this be related to https://github.com/OpenSC/OpenSC/issues/1295?

**Comment by dengert** [comment]
Get an opensc-debug log.  A SPY trace might also be helpful.  See https://github.com/OpenSC/OpenSC/wiki/Using-OpenSC

**Comment by francois4224** [comment]
here is the debug

```
OPENSC_DEBUG=9 pkcs15-tool --read-ssh 2 2> pkcs15
OPENSC_DEBUG=3  ssh -I "/Library/OpenSC/lib/pkcs11/onepin-opensc-pkcs11.so" ff@192.168.3.234 2> ssh
```
[ssh](https://github.com/OpenSC/OpenSC/files/2201550/ssh.txt)
[pkcs15](https://github.com/OpenSC/OpenSC/files/2201551/pkcs15.txt)

**Comment by dengert** [comment]
I could be all wrong, as I deal mostly with emulated cards with no  PKCS\#15 files.  Maybe someone else can have a look at this too.

But the best I can tell is the card has 2 RSA keys, 4 certificates and no public keys. (The public key can be found in the certificate. and some card drivers present it as a seperate object.)  

It also looks like the certificates are protected by the PIN.  PKCS11 will not see them unless logged in. 
Certificates are usually public objects. 

So this might be a issue with the initialization of the PKCS\#15 files to mark certificates private, but may not  be enforced by the card.  

Run `opensc-tool -f` that may show for detail of the files on the card. 

Also try `pkcs11-tool --test --login`  to see what it can do.

**Comment by francois4224** [comment]
You're right, on Yubico Neo and Feitian, I can find "Public RSA Key" data, but not on BEID.
However I can obtain the public key easily with "pkcs15-tool --read-ssh 2" without PIN, I imagine that it's extracted from the certificate:

```
$ opensc-tool -f
Using reader with a card: Generic Smart Card Reader Interface
3f00 type: DF, size: 65535
select[N/A] lock[N/A] delete[N/A] create[N/A] rehab[N/A] inval[N/A] list[N/A] 

sc_list_files() failed: Not supported
```
```
$ pkcs11-tool --test --login
Using slot 0 with a present token (0x0)
Logging in to "Basic PIN (BELPIC)".
Please enter User PIN: 
C_SeedRandom() and C_GenerateRandom():
  seeding (C_SeedRandom) not supported
  seems to be OK
Digests:
  all 4 digest functions seem to work
  MD5: OK
  SHA-1: OK
  RIPEMD160: OK
Signatures (currently only for RSA)
  testing key 0 (Authentication) 
  all 4 signature functions seem to work
  testing signature mechanisms:
    RSA-PKCS: OK
    SHA1-RSA-PKCS: OK
    MD5-RSA-PKCS: OK
    RIPEMD160-RSA-PKCS: OK
    SHA256-RSA-PKCS: OK
  testing key 1 (2048 bits, label=Signature) with 1 signature mechanism
Logging in to "Basic PIN (BELPIC)".
Please enter context specific PIN: 
error: PKCS11 function C_Sign failed: rv = CKR_FUNCTION_NOT_SUPPORTED (0x54)
Aborting.
    RSA-PKCS: 
```
```
$ pkcs15-tool -D
...
PIN [Basic PIN]
	Object Flags   : [0x3], private, modifiable
	ID             : 01
	Flags          : [0x30], initialized, needs-padding
	Length         : min_len:4, max_len:12, stored_len:8
	Pad char       : 0xFF
	Reference      : 1 (0x01)
	Type           : bcd
	Path           : 3f00

Private RSA Key [Authentication]
	Object Flags   : [0x3], private, modifiable
	Usage          : [0x4], sign
	Access Flags   : [0x1D], sensitive, alwaysSensitive, neverExtract, local
	ModLength      : 2048
	Key ref        : 130 (0x82)
	Native         : yes
	Path           : 3f00df00
	Auth ID        : 01
	ID             : 02
	MD:guid        : 73ad9bee-6bf4-5148-1d43-c102fbc1d446
...
X.509 Certificate [Authentication]
	Object Flags   : [0x3], private, modifiable
	Authority      : no
	Path           : 3f00df005038
	ID             : 02
	Encoded serial : 02 10 1000000000002227BD0C2D804DA9630C
...
```

**Comment by dengert** [comment]
The fact that the pkcs11-tool --test --login fails when trying to use the Signature key looks like it is related to  https://github.com/OpenSC/OpenSC/issues/1295 A debug log would show this, Note debug logs may show your PIN of other sensitive data, so you may want edit the debug log. Look for your pin  in the outgoing APDU as something like `3x 3x 3x 3x 3x FF FF FF` where xxxx are the pin digits.

Note that the pkcs11-tool --login does a login before looking for keys or certificates. But it looks like the ssh code does not. And it looks like it quits because it can not find a certificate or public key. (a SPY trace would show what ssh is doing.) The version of ssh may also make a difference.

**Comment by dengert** [comment]
You could try this patch. It compiles but not tested. It says **all** PKCS\#15 certificates are to be treated as PUBLIC.  So it is not a general fix and does not address #1293.

```
diff --git a/src/libopensc/pkcs15-cert.c b/src/libopensc/pkcs15-cert.c
index 6ee3e48..c4fb5d0 100644
--- a/src/libopensc/pkcs15-cert.c
+++ b/src/libopensc/pkcs15-cert.c
@@ -491,6 +491,9 @@ sc_pkcs15_decode_cdf_entry(struct sc_pkcs15_card *p15card, struct sc_pkcs15_obje
        if (obj->data == NULL)
                LOG_FUNC_RETURN(ctx, SC_ERROR_OUT_OF_MEMORY);
        memcpy(obj->data, &info, sizeof(info));
+#if 1
+       obj->flags &= ~SC_PKCS15_CO_FLAG_PRIVATE;
+#endif
 
        return 0;
 }
```
I expect ssh will work with the Authentication key but  fail with the  Signature key.  Because `pkcs11-tool -test --login` worked with the Authentication key but failed with the Signature Key. (#1293) An opensc debug log would show if your card has the same problem.

**Comment by francois4224** [comment]
Thanks, the patch solves my problem with ssh,

Is there any chance to integrate this in opensc, even if it's optional?

FYI, with the feitian key I have also this, but I don't want to use the signing part:
```
$ bin/pkcs11-tool --test --login
Using slot 0 with a present token (0x0)
Logging in to "UserPIN (GIDS card)".
Please enter User PIN: 
C_SeedRandom() and C_GenerateRandom():
  seeding (C_SeedRandom) not supported
  ERR: C_GenerateRandom failed: CKR_FUNCTION_NOT_SUPPORTED (0x54)
Digests: not implemented
Signatures (currently only for RSA)
  testing key 0 (RSA) 
error: PKCS11 function C_SignFinal failed: rv = CKR_GENERAL_ERROR (0x5)
Aborting.
```

**Comment by dengert** [comment]
There are two separate problems with the Belgium ID card. #1293 and #1295 
In https://github.com/OpenSC/OpenSC/issues/1430#issuecomment-405679886 I said it did not fix #1293, but it does. but it does not fix #1295 which deals with order of APDU commands sent to the card. 

@francois4224  asked: "Is there any chance to integrate this in opensc, even if it's optional?"
It's not up to me, but I would say yes.  No other card appears to have this problem. All other cards appear to not set the private bit or we would have heard about this same problem on other cards. 
 
But the issuers of the Belgium ID card may be setting a policy for some reason such as sensitive data in the certificates. But the card you are using does not enforce the policy. They may change this in future cards to enforce it by requiring a verify before reading the certificates. 

I would suggest the users of the  Belgium ID card @bigon  and @francois4224 and others need to research this with their card issuer. (They may stop setting the private bit if that was not intended.)

If they do start to enforce the policy in future cards, OpenSC will have a different problem. 
OpenSC only checks the private bit in PKCS\#11 but not at the PKCS\#15 layer or PKCS#15 tools.


------------------------

Further analysis shows the underlying problem is caused by the PKCS\#15 CommonObjectAttributes,  CommonObjectFlags, private flag being set to 1.  In both [pkc15](https://github.com/OpenSC/OpenSC/files/2201551/pkcs15.txt)  and [ssh](https://github.com/OpenSC/OpenSC/files/2201550/ssh.txt) the following is read from the card:
```
30 2C 30 17 0C 0E 41 75 74 68 65 6E 74 69 63 61 0,0...Authentica
74 69 6F 6E 03 02 06 C0 04 01 01 30 03 04 01 02 tion...À...0....
A1 0C 30 0A 30 08 04 06 3F 00 DF 00 50 38 30 27 ¡.0.0...?.ß.P80'
30 12 0C 09 53 69 67 6E 61 74 75 72 65 03 02 06 0...Signature...
C0 04 01 01 30 03 04 01 03 A1 0C 30 0A 30 08 04 À...0....¡.0.0..
06 3F 00 DF 00 50 39 30 23 30 0B 0C 02 43 41 03 .?.ß.P90#0...CA.
02 06 C0 04 01 01 30 06 04 01 04 01 01 FF A1 0C ..À...0......ÿ¡.
30 0A 30 08 04 06 3F 00 DF 00 50 3A 30 25 30 0D 0.0...?.ß.P:0%0.
0C 04 52 6F 6F 74 03 02 06 C0 04 01 01 30 06 04 ..Root...À...0..
01 06 01 01 FF A1 0C 30 0A 30 08 04 06 3F 00 DF ....ÿ¡.0.0...?.ß
00 50 3B 90 00
```
The debug levels set in the two logs are different and only  [pkcs15](https://github.com/OpenSC/OpenSC/files/2201551/pkcs15.txt)  shows the ASN1 parsing 
of the commonObjectAttributes -  flags(CommonObjectFlags) at line 965:
```
0x7fff95ff8380 14:15:34.885 [pkcs15-tool] asn1.c:1602:asn1_decode: Looking for 'flags', tag 0x3, OPTIONAL
0x7fff95ff8380 14:15:34.885 [pkcs15-tool] asn1.c:1393:asn1_decode_entry:   decoding 'flags', raw data:06C0
```
This is a bit string with 6 trailing unused bits, so "C0" is really only 2 bits, private and modifiable. both set to "1"
